src/main/java/io/reactivex/processors/AsyncProcessor.java

```diff
@@ -28,150 +26,244 @@
  * <p>The implementation of onXXX methods are technically thread-safe but non-serialized calls
  * to them may lead to undefined state in the currently subscribed Subscribers.
  * 
- * <p>Due to the nature Observables are constructed, the AsyncSubject can't be instantiated through
- * {@code new} but must be created via the {@link #create()} method.
- *
  * @param <T> the value type
  */
 public final class AsyncProcessor<T> extends FlowableProcessor<T> {
-    /** The state holding onto the latest value or error and the array of subscribers. */
-    final State<T> state;
-    /** 
-     * Indicates the subject has been terminated. It is checked in the onXXX methods in
-     * a relaxed matter: concurrent calls may not properly see it (which shouldn't happen if
-     * the reactive-streams contract is held).
-     */
-    boolean done;
+
+    @SuppressWarnings("rawtypes")
+    static final AsyncSubscription[] EMPTY = new AsyncSubscription[0];
+
+    @SuppressWarnings("rawtypes")
+    static final AsyncSubscription[] TERMINATED = new AsyncSubscription[0];
+
+    final AtomicReference<AsyncSubscription<T>[]> subscribers;
+    
+    /** Write before updating subscribers, read after reading subscribers as TERMINATED. */
+    Throwable error;
+    
+    /** Write before updating subscribers, read after reading subscribers as TERMINATED. */
+    T value;
     
     /**
-     * Constructs an empty AsyncSubject.
-     * @param <T> the observed and observable value type
-     * @return the new AsyncSubject instance.
+     * Creates a new AsyncProcessor.
+     * @param <T> the value type to be received and emitted
+     * @return the new AsyncProcessor instance
      */
     public static <T> AsyncProcessor<T> create() {
         return new AsyncProcessor<T>();
     }
     
-    protected AsyncProcessor() {
-        this.state = new State<T>();
+    /**
+     * Constructs an AsyncProcessor.
+     * @since 2.0
+     */
+    @SuppressWarnings("unchecked")
+    public AsyncProcessor() {
```

Comment:
Ugh, yeah. I don't really have any good suggestions here for that. The bounded constructor could be `int capacityHint, int size` but that's not a very descriptive for the caller. Are there any other examples that clash?

Personally I prefer the named static methods. They make it more clear as to what the values actually mean. Seeing `new ReplaySubject(16, 16)` doesn't tell me much where `ReplaySubject.createWithSize(16)` is much more explicit. Static methods are used for creating stream types (aside from subclassing) so I think it plays nicely.

> Should we `@Deprecate` what we don't need in the RC2 and then remove them RC3 or just delete stuff and mention it in the release log?

It would be nice to do the `@Deprecate` and remove where it doesn't impede the changes that need to be made in the library. Behavior changes may already be breaking consumers so minimizing the amount of work that needs to be done immediately when you're updating the version is helpful.

I would definitely call them out in the change log regardless.